### PR TITLE
Enforced client side validation by default.

### PR DESCRIPTION
**Pull request by quigamdev**
Issue description: 
When a string value is in the textbox which was bound to nullable numeric type then validation pass. 
This is wrong behavior. Validation has to fail and a user should be notified that value that is in the textbox has an incorrect format.

Solution: 
I have enforced DotvvmEnforceClientFormat on all numeric nullable types. Due to compatibility, I have just added this attribute to these properties and added Enforce property to be able to disable it.

### CODE REVIEW

File: src/DotVVM.Core/ComponentModel/DataAnnotations/ClientFormatAttribute.cs [hunk 1]
```diff
@@ -9,15 +9,24 @@ namespace System.ComponentModel.DataAnnotations
     /// Checks whether the value matches the format specified by DotVVM control (e.g. TextBox).
     /// The value can be empty string or null. If it is not, it must match the FormatString that the control specifies.
     /// </summary>
+    [Obsolete("The client format is enforced by default. If you need to disable it use ClientFormatAttribute", true)]
```
Comment (exyi) [inline src/DotVVM.Core/ComponentModel/DataAnnotations/ClientFormatAttribute.cs]: I'd also include a note that `ClientFormatAttribute` also allows you to configure the behavior.

File: src/DotVVM.Framework/ViewModel/Validation/ViewModelValidationRuleTranslator.cs [hunk 2]
```diff
@@ -34,7 +36,9 @@ public virtual IEnumerable<ViewModelPropertyValidationRule> TranslateValidationR
                         validationRule.ClientRuleName = "range";
                         validationRule.Parameters = new[] { rangeAttr.Minimum, rangeAttr.Maximum };
                         break;
-                    case DotvvmEnforceClientFormatAttribute enforceClientFormatAttr:
+                    case DotvvmClientFormatAttribute enforceClientFormatAttr:
+                        addEnforceClientFormat = false;
+                        if (enforceClientFormatAttr.Disable) yield break;
```
Comment (exyi) [inline src/DotVVM.Framework/ViewModel/Validation/ViewModelValidationRuleTranslator.cs]: Hey, this `yield break;` interrupts the execution and will suppress any following validators.

File: src/DotVVM.Framework/ViewModel/Validation/ViewModelValidationRuleTranslator.cs [hunk 3]
```diff
@@ -46,9 +50,21 @@ public virtual IEnumerable<ViewModelPropertyValidationRule> TranslateValidationR
                         validationRule.ClientRuleName = string.Empty;
                         break;
                 }
-      
+
                 yield return validationRule;
             }
+            // enforce client format by default
+            if (addEnforceClientFormat && property.PropertyType.IsNullable() && Nullable.GetUnderlyingType(property.PropertyType).IsNumericType())
```
Comment (exyi) [inline src/DotVVM.Framework/ViewModel/Validation/ViewModelValidationRuleTranslator.cs]: You should also add it to nullable `DateTime`, not only numerics

File: src/DotVVM.Samples.Common/ViewModels/FeatureSamples/Validation/EnforceClientSideValidationDisabledViewModel.cs [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+using System.Linq;
+using System.Text;
+using DotVVM.Framework.ViewModel;
+using System.ComponentModel.DataAnnotations;
+
+namespace DotVVM.Samples.BasicSamples.ViewModels.FeatureSamples.Validation
+{
+	public class EnforceClientSideValidationDisabledViewModel
+    {
+        [DotvvmClientFormat(Disable = true, ErrorMessage = "This message will not be display. Because this client format enforcing is disabled.")]
```
Comment (exyi) [inline src/DotVVM.Samples.Common/ViewModels/FeatureSamples/Validation/EnforceClientSideValidationDisabledViewModel.cs]: display**ed**

**Comment by exyi** [comment]
Note that this PR includes quite nasty breaking changes, specifically when viewmodel includes a nullable numeric property it's validation is enforced, which was not the case previously. We should make sure that we inform everyone to verify that nothing breaks in their applications.

File: src/DotVVM.Framework/ViewModel/Validation/ViewModelValidationRuleTranslator.cs (hunk 2 above)
Comment (quigamdev) [reply src/DotVVM.Framework/ViewModel/Validation/ViewModelValidationRuleTranslator.cs]: Oh, that's a silly mistake. Sorry about that.

**Comment by quigamdev** [comment]
I have made some additional changes. @exyi Please, could you review the changes?

File: src/DotVVM.Framework/ViewModel/Validation/ViewModelValidationRuleTranslator.cs [hunk 5]
```diff
@@ -34,10 +36,16 @@ public virtual IEnumerable<ViewModelPropertyValidationRule> TranslateValidationR
                         validationRule.ClientRuleName = "range";
                         validationRule.Parameters = new[] { rangeAttr.Minimum, rangeAttr.Maximum };
                         break;
-                    case DotvvmEnforceClientFormatAttribute enforceClientFormatAttr:
+                    case DotvvmClientFormatAttribute enforceClientFormatAttr:
+                        addEnforceClientFormat = false;
+                        if (enforceClientFormatAttr.Disable)
+                            break;
```
[the remark this answers is not in the document]
Comment (quigamdev) [reply src/DotVVM.Framework/ViewModel/Validation/ViewModelValidationRuleTranslator.cs]: When ECFAttribute.Disable is false, it means that the whole attribute should be skipped.

**Pull request merged**